c++/src/kj/string.h

```diff
@@ -554,6 +554,107 @@ template <typename T, typename = decltype(toCharSequence(instance<T&>()))>
+
+template <typename T>
+class HasCappedCapacity_ {
+  using yes = char;
+  using no = yes[2];
+
+  template<typename C> static auto test(void*) -> decltype(C::capacity(), yes{});
+  template<typename C> static no& test(...);
+
+public:
+  static constexpr bool value = sizeof(test<T>(nullptr)) == sizeof(yes);
+};
+template <typename T>
+class HasCappedCapacity_<SignalSafeCharSequence<T>>: public HasCappedCapacity_<T> {};
+template <typename T, size_t N>
+class HasCappedCapacity_<T(&)[N]> { static constexpr bool value = true; };
+template <typename T, size_t N>
+class HasCappedCapacity_<const T(&)[N]> { static constexpr bool value = true; };
+
+template <typename T> static inline constexpr bool hasCappedCapacity() {
+  // This returns true if T::capacity() is a method that exists, false otherwise. This is a
+  // helper function for isToCharSequenceResultingInCappedCapacity as I couldn't figure out how to
+  // make the SFINAE magic less verbose.
+  return HasCappedCapacity_<T>::value;
+}
+
+template <typename T>
+struct IsToCharSequenceResultingInCappedCapacity_ {
+  static constexpr bool value = hasCappedCapacity<decltype(tryToCharSequence(instance<T*>()))>();
+};
+template <typename T>
+struct IsToCharSequenceResultingInCappedCapacity_<T&> {
+  static constexpr bool value = IsToCharSequenceResultingInCappedCapacity_<T>::value;
+};
+template <typename T>
+struct IsToCharSequenceResultingInCappedCapacity_<const T&> {
+  static constexpr bool value = IsToCharSequenceResultingInCappedCapacity_<T>::value;
+};
+template <typename T>
+struct IsToCharSequenceResultingInCappedCapacity_<T&&> {
+  static constexpr bool value = IsToCharSequenceResultingInCappedCapacity_<T>::value;
+};
+template <typename T>
+struct IsToCharSequenceResultingInCappedCapacity_<const T&&> {
+  static constexpr bool value = IsToCharSequenceResultingInCappedCapacity_<T>::value;
+};
+template <typename T>
+static constexpr inline bool isToCharSequenceResultingInCappedCapacity() {
+  // Returns true if toCharSequence<T> returns a stringifiable type that has an upper-bound capacity
+  // requirement known at compile-time (i.e. CappedArray/FixedArray). This is primarily useful for
+  // indicating whether composite types are safe to stringify in a signal handler context as
+  // otherwise you would need to allocate memory. There's an alternate universe where KJ_STRINGIFY
+  // could optionally accept the buffer to append into rather than having to return a type, but
+  // that seems overly complex.
+  return IsToCharSequenceResultingInCappedCapacity_<T>::value;
+};
+
+template <typename T> struct CapacityHelper_ { static constexpr size_t value = T::capacity(); };
+template <typename T>
+struct CapacityHelper_<SignalSafeCharSequence<T>> {
+  static constexpr size_t value = T::capacity();
+};
+template <typename T>
+struct CappedCapacityOfToCharSequence_ {
+  static constexpr size_t capacity = CapacityHelper_<
+      decltype(toCharSequence(instance<T>()))>::value;
```

Comment:
I only used `tryToCharSequence` because of SFINAE failures being annoying to work out. Now that I have the `VoidSfinae` trick worked out I've changed it to use `toCharSequence` consistently. `tryToCharSequence` is unimportant here because it returns `kj::StringPtr` on failure which doesn't have a capacity & otherwise just invokes `toCharSequence`.